---
title: "assignment"
author: "Rebecca Waag"
date: "`r Sys.Date()`"
output: html_document
---

# Week11: DNA methylation and CpG islands

## Exercise11: DNAme visualization

1. Plot a heatmap of the methylation levels of the genes in top 5 DMR regions
2. Run an enrichment analysis (rGREAT) of the genes within DMRs vs all genes of chromosome 22. 
-->Describe what your enrichment analysis is testing exactly & describe the results.

```{r}
suppressPackageStartupMessages({
  library(GenomicRanges)
  library(epiwraps)
  library(ggplot2)
  library(rGREAT)
  library(AnnotationHub)
  library(ensembldb)
  library(bsseq)
  library(BiocParallel)
  library(edgeR)
  library(DMRcate)
  library(rtracklayer)
  library(sechm)
  library(pheatmap)
  library(viridis)
  library(data.table)
})

set.seed(40)
```

### Load the data

```{r, eval=FALSE}
#download data
setwd("C:/Users/rebec/Desktop/w11")
options(timeout = 6000)
download.file("https://ethz-ins.org/content/w11_practical.zip", "w11_practical.zip")
unzip("w11_practical.zip")
```

```{r}
#read in file
setwd("C:/Users/rebec/Desktop/w11")
bs <- readRDS("bs.rds")
#rowRanges(bs) #578097 ranges, only chr22
#pData(bs) #6 samples with 2 types and 3 pairs
#bs #578097 cpg sites, 6 samples
```

### Perform differential methylation testing

Perform differential methylation testing to reveal differentially methylated regions between control and cancer samples, take pair data into account

```{r}
#design modelmatrix to find significantly altered regions between control and cancer, take pair data into account
design <- model.matrix(~Type+Pair, data=pData(bs)) 
#design

#construct design matrix for edgeR analysis of methylation count data
methdesign <- modelMatrixMeth(design)
#methdesign

#annotate bisulfite sequencing with probe weights and chromosomal position
seqAnnot <- sequencing.annotate(bs, methdesign, all.cov=TRUE, coef="Typenormal")

#identify differentially methylated regions (DMR)
dmrcateRes <- dmrcate(seqAnnot, C=2, min.cpgs = 10, pcutoff=0.01)
dmrcateRes #calculates 2193 DMRs

#create a GRanges object from dmrcate output
dmrRanges <- extractRanges(dmrcateRes, genome="hg38")
#dmrRanges #Granges with 2193 ranges and 8 meta columns
#DMrs are ranked by Fisher's multiple comparison statistics

#save as rds object
saveRDS(dmrRanges, "dmr.rds")

#remove all ranges that have NA overlapping.genes
dmrRangesGenes <- dmrRanges[!is.na(dmrRanges$overlapping.genes)]
dmrRangesGenes #1632 differentially methylated regions
```

1632 differentially methylated regions (DMRs) with overlapping genes are identified between control and cancer samples, taking available pair data into account. The top differentialy methylated region (based on Fisher stats) is associated with the ZNRF3 locus.

```{r}
# genes
ah <- AnnotationHub(localHub=TRUE)
ensdb <- ah[["AH109336"]]

#extract genes on chrom22
chr22 <-  GRanges(seqnames=Rle(c("22")), 
                  ranges = IRanges(1, end=195471971))
genesChr22 <- genes(ensdb, columns=c("gene_seq_start", "gene_seq_end", "gene_name"),
                    filter=GRangesFilter(chr22))
seqlevelsStyle(genesChr22) <- "UCSC"
```

```{r}
#obtain coordinates of the genes within the top 5 DMRs
#order DMRs based on min_smoothed_fdr and obtain first 5 entries
topIdx <- order(dmrRangesGenes$min_smoothed_fdr)[1:5]
topIdx #[1] 1318 1447 1497 1508  402
dmrRangesGenes[c(1318, 1447, 1497, 1508, 402),] #two of the DMRs have 2 overlapping genes, the other three have one overlapping gene

#split the overlaping gene names to obtain all entries separately and remove all NA entries
genesDmr <- unlist(tstrsplit(dmrRangesGenes[topIdx]$overlapping.genes, split=", "))
genesDmr <- genesDmr[!is.na(genesDmr)]
genesDmr #7 genes are available, but one seems arbitrary

#only keep genes that are present in the list of genes of Chr22
dmrGenes <- genesChr22[genesChr22$gene_name %in% genesDmr]
dmrGenes #the top 5 DMR contain 6 genes

#obtain gene names within top 5 DMR
dmrGenes$gene_name
```

The six genes CDC45, SHISAL1, FBLN1, PKDREJ, TRMU and CELSR1 are within the top 5 DMRs. 

### Plot heatmap of methylation levels of genes in top 5 DMR regions

```{r}
#use getMeth to obtain methylation estimates for BSseq objects of genes on chromosome 22
metPr <- bsseq::getMeth(bs, 
                         regions = genesChr22, #dataframe or GRanges specifying genomic regions
                         what="perRegion") #type of return object
#metPr #returns a DelayedMatrix containing average methylation levels of all genes on chrom22
colnames(metPr) <- colnames(bs) #add sample names to columns
rownames(metPr) <- genesChr22$gene_name #add gene names to rows
#dim(metPr) #contains 1482 regions
metPr <- metPr[!is.na(rowSums(metPr)),] #remove all regions with average methylation level NA
#dim(metPr) #contains 1365 regions

annotationCol <- as.data.frame(pData(bs)[,c("Type", "Pair")])
annotationCol
rownames(annotationCol) <- colnames(metPr)

#extract the top genes
#metPr_top5<-metPr[c("CDC45", "SHISAL1", "FBLN1", "PKDREJ", "TRMU", "CELSR1"),]
metPr_top5<-metPr[dmrGenes$gene_name,]
metPr_top5
pheatmap::pheatmap(metPr_top5, 
                   cluster_rows=TRUE,
                   cluster_cols=FALSE,
                   annotation_col=annotationCol,
                   show_rownames = TRUE,
                   color=rocket(10))
```

Clear differences between control and cancer samples can be observed for genes, such as TRMU which indicates hypomethylation of this gene in cancer conditions. For other genes, the differences are not easy to interpret based on only this heatmap and indicate that the genes themselves might not contribute solely to the differences observed on the level of the DMRs between the groups. 

### Enrichment analysis of genes within DMRs vs. all genes (chrom22)

Enrichment analysis is performed on ALL genes that are within DMRs (not only the top 5 DMRs as 6 genes are not enough for enrichment analysis). All genes present on chrom22 serve as the background. 

```{r}
#obtain coordinates of the genes within DMRs
#split the overlaping gene names to obtain all entries separately and remove all NA entries
genesDmr_all <- unlist(tstrsplit(dmrRangesGenes$overlapping.genes, split=", "))
genesDmr_all <- genesDmr_all[!is.na(genesDmr_all)]
#length(genesDmr_all) #2002 genes

#only keep genes that are present in the list of genes of Chr22
dmrGenes_all <- genesChr22[genesChr22$gene_name %in% genesDmr_all]
dmrGenes_all #532 genes are within all DMRs
```

532 genes are within all DMRs on chrom22. 

```{r}
#perform enrichment analysis
#send data to the GREAT server to run an enrichment analysis
job_1 <- rGREAT::submitGreatJob(gr=dmrGenes_all, bg=genesChr22, species="hg38")
res_1 <- rGREAT::getEnrichmentTables(job_1)
#names(res_1)

#plot the top 10 Biological processes
bp_1 <- res_1$`GO Biological Process`
ggplot(head(bp_1,10), aes(Hyper_Fold_Enrichment, reorder(name, Hyper_Adjp_BH), size=Hyper_Foreground_Region_Hits, color=-log10(Hyper_Adjp_BH))) + geom_point() + scale_color_viridis_c() + ggtitle("Genes within DMRs: Biological Processes")

```

This enrichment analysis tests, whether those genes located within differentially methylated regions (DMRs) on chrom22 show an enrichment for any biological processes compared to ALL genes that are located on the same chromosome. 
The most striking enrichment seems to be "positive regulation of cell adhesion" which shows a particularly strong fold change, meaning that genes within DMRs seem to be especially involved in this process. Signaling in general seems to be an important factor, as several processes are enriched (regulation of signaling, Cell communication, cell adhesion, signal transduction).