---
title: "Marrocco_Peak-Quantification"
author: "Rebecca Waag"
date: "8 6 2023"
output:
  html_document: default
  pdf_document: default
---

# Quantification of ATAC data

Perform differential expression analysis, to define which peaks are different between the individual groups.
The Marrocco dataset contains 4 groups of animals:
1 - Control
2 - ARS : animals that underwent restraint stress (ARS) and were euthanized 1 week later
3 - FSS : animals that underwent forced swim stress (FSS) for 6 min and were euthanized 40 minutes later
4 - Doublehit: animals that underwent ARS and one week later 6 min FSS. These animals were euthanized 40 minutes after the FSS. 

We will focus on identifying those regions that change their accessibility in response to an acute stress session. 

```{r}
suppressPackageStartupMessages({
  library(epiwraps)
  library(motifmatchr)
  library(MotifDb)
  library(universalmotif)
  library(sechm)
  library(BiocParallel)
  library(ggplot2)
  library(chromVAR) # for motif accessibility estimation
  library(SummarizedExperiment) # data structure
  library(edgeR)
  library(stringr) #for word function
  library(rtracklayer)
  library(scales)
})
# to control multithreading, unix users can use:
register(MulticoreParam(4))
```

## Perform Differential analysis on peaks

### Import peaks

Prepare a Granges object containing all peaks that appear in one of the samples. Here we are working on all peaks from the ATAC dataset (not just NF-peaks)

```{r}
#obtain peak names
peaksM <- list.files("/mnt/bohacek/rebecca/training/EpiC/Project/data_Marrocco/peaks/", pattern = "narrowPeak", full = TRUE)
#peaksM

#rename peaks
names(peaksM) <- gsub("\\.narrowPeak","",basename(peaksM))

#import peaks
peaksM <- lapply(peaksM, FUN=rtracklayer::import)
length(peaksM) #24 peak files
```

### Getting a consensus set of windows

A consensus set of peaks, containing all peaks that appear in at least one of the samples is produced. Blacklisted regions, those that generally have high signals independent of the experiment, are removed. Of the remaining peaks, only those that appear in at least three samples are kept. 

```{r}
#create consensus peaks
merged_peaksM <- reduce(unlist(GRangesList(peaksM)), with.revmap=TRUE)
#length(merged_peaksM) #196887

#remove blacklisted regions from consensus peaks
blacklist <- import("/reference/Mus_musculus/GRCm38.blacklist.bed")
merged_peaksM <- merged_peaksM[!overlapsAny(merged_peaksM, blacklist)]
#length(merged_peaksM) #196815

#only keep peaks that appear in at least 3 samples
merged_peaksM <- merged_peaksM[lengths(merged_peaksM$revmap)>=3]
#length(merged_peaksM) #99852

```

### Getting the counts per peak

Prepare a matrix of fragment counts in the merged peaks as a summarized experiment object for DEA. 

```{r}
#list bam files
bams <- list.files("/mnt/bohacek/Marrocco_GSE200670_stressATAC/aligned/", pattern="bam$", full=TRUE)

#create summarized experiment object
#se <- chromVAR::getCounts(alignment_files=bams, 
#                             peaks=merged_peaksM, 
#                             paired =  TRUE, 
#                             format = "bam")
se <- readRDS("/mnt/bohacek/rebecca/training/EpiC/Project/Marroccopeaks.SE.rds")

#explore assay
se
assays(se) #contains counts

#plot sequencing depth
#barplot(se$depth)
#sequencing depth of all samples is pretty different

#adjust the colnames
#colnames(se)
colnames(se) <- gsub("\\.bam","",basename(bams))

#extract group from filename and add as metadata
se$group <- word(colnames(se), 1, sep = "_")
se$group <- as.factor(se$group)
#se$group
#levels(se$group)
se$group<-relevel(se$group, "Control")
levels(se$group)

#extract sex from filename and add as metadata
se$sex<-word(colnames(se), 2, sep = "_")
se$sex <- as.factor(se$sex)
#se$sex
#levels(se$sex)
se$sex<-relevel(se$sex, "male")
levels(se$sex)

#extract ID from filename and add as metadata
se$id<-word(colnames(se), 3, sep = "_")
se$id <- as.factor(se$id)
se$id

#by default there are no rownames, us regions as row.names
row.names(se) <- as.character(granges(se))
```

### Data Exploration

Explore data and visualize differences in the region counts using PCA.

```{r, out.width=c('80%'), fig.align='center', fig.cap=c('FIGURE 1: PCA of peak counts for original groups')}
#plot PCA of experiment visualizing
colors <- c("blue", "darkgreen", "red", "black")
plotMDS(se, col=colors[se$group], labels=se$group, gene.selection="common")
legend("topleft", fill=colors, legend=levels(se$group), col=colors, ncol=2)
```

The groups do not nicely separate along PC1 or PC2 (Fig. 1), therefore, a surrogate variable is added to the analysis as a covariate.

### Differential analysis with edgeR

```{r}
#prepare DGE object
y <- DGEList(counts=assays(se)$counts)

#calculate normalization, assumptions are met in this experiment, so we can use TMM normalization
y <- calcNormFactors(y)

#filter by expression
#y <- y[filterByExpr(y,group=se$group),]

#calculate covariate
se <- SEtools::svacor(se, ~ sex + group, n.sv=1)
assays(se) #corrected assay added

#specify model matrix to compare between the different groups, correcting for sex and SV1
mm <- model.matrix(~ sex + SV1 + group, data=as.data.frame(colData(se)))
#we use an intercept here, meaning that we compare against the controls set above

#estimate dispersion
y <- estimateDisp(y,mm)
y$common.dispersion
#plot dispersion estimates
#plotBCV(y)

#fit model
fit <- glmQLFit(y, mm)

#check coefficients
colnames(mm)
#[1] "(Intercept)"     "sexfemale"       "SV1"             "groupARS"        "groupDouble-hit" "groupFSS"

#create list for results
results_M <- list()

#compare ARS vs. control
results_M[["Marrocco.ARS"]] <- glmQLFTest(fit, coef="groupARS")
topTags(results_M[["Marrocco.ARS"]])
summary(decideTests(results_M[["Marrocco.ARS"]]))
#       groupARS
#Down          0
#NotSig    99852
#Up            0

#compare FSS vs. control
results_M[["Marrocco.FSS"]] <- glmQLFTest(fit, coef="groupFSS")
topTags(results_M[["Marrocco.FSS"]])
summary(decideTests(results_M[["Marrocco.FSS"]]))
#       groupFSS
#Down          0
#NotSig    99818
#Up           34

#compare Doublehit vs. control
results_M[["Marrocco.Doublehit"]] <- glmQLFTest(fit, coef="groupDouble-hit")
topTags(results_M[["Marrocco.Doublehit"]])
summary(decideTests(results_M[["Marrocco.Doublehit"]]))
#      groupDouble-hit
#Down                 0
#NotSig           99817
#Up                  35

#calculate fdr
for(i in names(results_M)){
  results_M[[i]]$table$FDR <- p.adjust(results_M[[i]]$table$PValue, method = "fdr")
}

#calculate which regions are overlapping between FSS and Double-hit hits
length(Reduce(intersect, list(which(results_M$Marrocco.FSS$table$FDR<0.05),which(results_M$Marrocco.Doublehit$table$FDR<0.05)))) #24
```
The DA reveals 34 signficantly accessible regions between FSS and control. Between double-hit and control, there are 35 significantly accessible regions. 
ARS seems to do not elicit any differential accessible regions 1 week later, indicating that (restraint) stress has no long-lasting effect on chromatin accessibility. 

Of the differentially accessible regions of FSS and Double-hit, 24 are consistent between those two conditions, indicating a strong influence of the current stress exposure both groups of animals were exposed to. Therefore, new factors will be introduced to take into account the acute stress exposure, as well as the history of stress some animals underwent. The new factors are acute stress (FSS, Double-hit) and a history of stress (ARS, double-hit).

```{r}
#add new metadata based on the groups
se$group

#add history of stress factor (ARS)
se$history <- c(rep("history", 6), rep("no", 6), rep("history", 6), rep("no",6))
se$history <- as.factor(se$history)
se$history<-relevel(se$history, "no")
se$history

#add acute stress factor (FSS)
se$stress <- c(rep("no", 12), rep("acute", 12))
se$stress <- as.factor(se$stress)
se$stress<-relevel(se$stress, "no")
se$stress

#prepare DGE object
y <- DGEList(counts=assays(se)$counts)

#calculate normalization, assumptions are met in this experiment, so we can use TMM normalization
y <- calcNormFactors(y)

#filter by expression
#y <- y[filterByExpr(y,group=se$group),]

#calculate covariate
se <- SEtools::svacor(se, ~ sex + history + stress, n.sv=1)
assays(se) #corrected assay added

#specify model matrix to compare between the stress and control, correcting for history of stress, sex and SV1
mm <- model.matrix(~ sex + SV1 + history + stress, data=as.data.frame(colData(se)))
#we use an intercept here, meaning that we compare against the controls that were set before

#estimate dispersion
y <- estimateDisp(y,mm)

#fit model
fit <- glmQLFit(y, mm)

#check coefficients
colnames(mm)
#[1] "(Intercept)"    "sexfemale"      "SV1"            "historyhistory" "stressacute"

#get results
#results_M <- list()

#compare history vs. control
results_M[["Marrocco.history"]] <- glmQLFTest(fit, coef="historyhistory")
topTags(results_M[["Marrocco.history"]])
summary(decideTests(results_M[["Marrocco.history"]]))
#       groupARS
#Down          0
#NotSig    99852
#Up            0

#compare acute stress vs. control
results_M[["Marrocco.acute"]] <- glmQLFTest(fit, coef="stressacute")
topTags(results_M[["Marrocco.acute"]])
summary(decideTests(results_M[["Marrocco.acute"]]))
#       stressacute
#Down          1
#NotSig    99518
#Up          333

#calculate fdr
for(i in names(results_M)){
  results_M[[i]]$table$FDR <- p.adjust(results_M[[i]]$table$PValue, method = "fdr")
}
```

```{r, out.width=c('80%'), fig.align='center', fig.cap=c('FIGURE 2: Volcanoplot showing significantly accessible regions between control and acutely-stressed animals')}
#look at data using volcano plot to show regions with increased or decreased baccessibility of acute stress
#classify sign. up and downregulated peaks
results_M$Marrocco.acute$table$diff[results_M$Marrocco.acute$table$FDR < 0.05 & results_M$Marrocco.acute$table$logFC >0] <- "UP"
results_M$Marrocco.acute$table$diff[results_M$Marrocco.acute$table$FDR < 0.05 & results_M$Marrocco.acute$table$logFC <0] <- "DOWN"

#visualize results
ggplot(results_M$Marrocco.acute$table, aes(logFC, -log10(FDR), col=results_M$Marrocco.acute$table$diff)) + geom_point(size=1) + theme_bw() + scale_color_manual(values = c("blue", "red"), labels =c ("downregulated", "upregulated", "not significant")) + ggtitle('Significantly accessible regions: Acute stress') + labs(color = 'DEA, FDR<0.05', x = expression("logFC"), y = expression("-log"[10]*"p-value"))
```
Visualizing the regions in a volcano plot shows that all but one regions are significantly more open upon stress, meaning they show increased accessibility (Fig. 2). 

```{r, out.width=c('80%'), fig.align='center', fig.cap=c('FIGURE 3: Heatmap showing differentially-accessible regions between control and acutely-stressed animals (corrected)')}
#add metadata column to SE object to define colors
metadata(se)$anno_colors$stress <- c(no="#A8BED0", acute="#2F4E67")
metadata(se)$anno_colors$history <- c(no="#F7D9A3", history="#DB9413")
metadata(se)$anno_colors$sex <- c(male="#A87974", female="#E4A6A0")

#visualize uncorrected counts in heatmap
#sechm(se,features = rownames(topTags(results_M$Marrocco.acute, n = Inf, p.value = 0.05)), do.scale=TRUE, #assayName="counts", row_title="Stress-responsive regions", gaps_at = "stress", cluster_rows = TRUE, #top_annotation=c("sex", "history", "stress"))

#visualize corrected counts in heatmap
sechm(se,features = rownames(topTags(results_M$Marrocco.acute, n = Inf, p.value = 0.05)), do.scale=TRUE, assayName="corrected", row_title="Stress-responsive regions", gaps_at = "stress", cluster_rows = TRUE, top_annotation=c("sex", "history", "stress"))

#save summarized experiments object
saveRDS(se, file="Marroccopeaks.SE.rds")
```

334 regions are differentially accessible between acutely stressed (FSS, double-hit) and non-acutely stressed (Control, ARS) animals, correcting for the different history of stress, sex and covariate (Fig. 3). Most of these (333) are upregulated in response to acute stress. No differentially accessible regions are found when testing for the effect of a history of stress. 

```{r}
#subset all significant stress peaks from Marrocco data set
#dim(results_M[["Marrocco.acute"]]$table) #99852
#length(merged_peaksM) #99852
peaksM_stress_sig <- merged_peaksM[which(results_M[["Marrocco.acute"]]$table$FDR<0.05)]
#length(peaksM_stress_sig) #334 peaks

#subset sign. upregulated peaks
peaksM_stress_sigup <- merged_peaksM[which((results_M[["Marrocco.acute"]]$table$FDR<0.05) & (results_M[["Marrocco.acute"]]$table$logFC>0))]
#length(peaksM_stress_sigup) #333

#save peaks
saveRDS(peaksM_stress_sig, file="peaksM_stress_sig.rds")
saveRDS(peaksM_stress_sigup, file="peaksM_stress_sigup.rds")
```

### Enrichment analysis of stress-upregulated peaks

Enrichment analysis on the peaks was performed using the rGREAT server. As background, all accessible peaks (merged peaks file that was previously used to construct the SE object) from the Marrocco dataset will be used. 

```{r, out.width=c('100%'), fig.align='center', fig.cap=c('FIGURE 4: Enrichment analysis of significantly more accessible regions')}
#perform enrichment analysis using rGREAT server
library(rGREAT)

#adjust seqlevelStyle for rGREAT server
seqlevelsStyle(merged_peaksM) <- "UCSC"
seqlevelsStyle(peaksM_stress_sig) <- "UCSC"
seqlevelsStyle(peaksM_stress_sigup) <- "UCSC"

#only keep common chromosome names and clean up Granges for rGREAT
seqnames(merged_peaksM)
chrom <- c("chr1", "chr10", "chr11", "chr12", "chr13", "chr14", "chr15", "chr16", "chr17", "chr18", "chr19", "chr2", "chr3", "chr4", "chr5", "chr6", "chr7", "chr8", "chr9", "chrM", "chrX", "chrY")
merged_peaksM_chrom<- keepSeqlevels(merged_peaksM, value=chrom, pruning.mode="coarse")
length(merged_peaksM_chrom) #99764
names(merged_peaksM_chrom) <-NULL
seqnames(peaksM_stress_sig)
peaksM_stress_sig_chrom<- keepSeqlevels(peaksM_stress_sig, value=chrom, pruning.mode="coarse")
length(peaksM_stress_sig_chrom) #334
names(peaksM_stress_sig_chrom) <-NULL
peaksM_stress_sigup_chrom<- keepSeqlevels(peaksM_stress_sigup, value=chrom, pruning.mode="coarse")
length(peaksM_stress_sigup_chrom) #333
names(peaksM_stress_sigup_chrom) <-NULL

#get enrichment results for all sign peaks
#job_stress_sig <- rGREAT::submitGreatJob(gr=peaksM_stress_sig_chrom, bg=merged_peaksM_chrom, species="mm10") #we use mm10, cause GRCm38 is not supported
#res_stress_sig <- rGREAT::getEnrichmentTables(job_stress_sig)
#res_stress_sig

#plot the top 30 Biological processes
#bp_stress_sig <- res_stress_sig$`GO Biological Process`
#bp_stress_sig
#sum(bp_stress_sig$Hyper_Adjp_BH<0.05) #54 GO terms are sign. enriched
#ggplot(head(bp_stress_sig,30), aes(Hyper_Fold_Enrichment, reorder(name, Hyper_Adjp_BH), size=Hyper_Foreground_Region_Hits, color=-log10(Hyper_Adjp_BH))) + geom_point() + scale_color_viridis_c() + ggtitle("Stress-responsive peaks: Biological Processes") + scale_y_discrete(labels = label_wrap(50))

#get enrichment results for all sign upregulated peaks
job_stress_sigup <- rGREAT::submitGreatJob(gr=peaksM_stress_sigup_chrom, bg=merged_peaksM_chrom, species="mm10") #we use mm10, cause GRCm38 is not supported
res_stress_sigup <- rGREAT::getEnrichmentTables(job_stress_sigup)
#res_stress_sig

#plot the top 30 Biological processes
bp_stress_sigup <- res_stress_sigup$`GO Biological Process`
#bp_stress_sig
sum(bp_stress_sigup$Hyper_Adjp_BH<0.05) #52 GO terms are sign. enriched
ggplot(head(bp_stress_sigup,30), aes(Hyper_Fold_Enrichment, reorder(name, Hyper_Adjp_BH), size=Hyper_Foreground_Region_Hits, color=-log10(Hyper_Adjp_BH))) + geom_point() + scale_color_viridis_c() + ggtitle("Stress-responsive peaks: Biological Processes") + scale_y_discrete(labels = label_wrap(50))
```

Performing enrichment analysis on all significantly more accessible regions between acutely stressed and control animals in the Marrocco dataset, reveals enrichment of IRE1-mediated unfolded protein response and kinase activity, which are known to be involved in the general stress response (e.g. IRE1 in ER stress) (Fig. 4). Interestingly, the response to glucocoticoids is strongly enriched, speaking to the important role of glucocorticoids in the stress response. 
As almost all sign. differentially accessible regions are upregulated (with only one region downregulated), performing enrichment analysis on all differentially accesible regions gives very similar results. 

## Perform Differential analysis on NF-peaks

So far, analysis was performed on ALL accessible regions of the genome, in a next step, the same analysis will be performed using only the peaks (==NF-peaks) called on the nucleosome-free (NF) bams. These regions should be especially relevant for TF-guided action.

## Import NF peaks only

```{r}
#obtain peak names
peaksM_NF <- list.files("/mnt/bohacek/rebecca/training/EpiC/Project/data_Marrocco/NF-peaks/", pattern = "narrowPeak", full = TRUE)
#peaksM_NF

#rename peaks
names(peaksM_NF) <- gsub("\\.narrowPeak","",basename(peaksM_NF))

#import peaks
peaksM_NF <- lapply(peaksM_NF, FUN=rtracklayer::import)
#length(peaksM_NF) #25 peak files

#get rid of merged peaks
peaksM_NF <- peaksM_NF[1:24]
#length(peaksM_NF) #24 files
```

### Getting a consensus set of windows

```{r}
#create consensus peaks
merged_peaksM_NF <- reduce(unlist(GRangesList(peaksM_NF)), with.revmap=TRUE)
length(merged_peaksM_NF) #557480

#remove blacklisted regions from consensus peaks
blacklist <- import("/reference/Mus_musculus/GRCm38.blacklist.bed")
merged_peaksM_NF <- merged_peaksM_NF[!overlapsAny(merged_peaksM_NF, blacklist)]
length(merged_peaksM_NF) #557470

#only keep peaks that appear in at least 3 samples
merged_peaksM_NF <- merged_peaksM_NF[lengths(merged_peaksM_NF$revmap)>=3]
length(merged_peaksM_NF) #141235
#several of these peaks are very specific to samples and are not present across several of these
```

### Getting the counts per peak

We prepare another se object that contains the NF-peaks. 

```{r}
#list bam files
bams_NF <- list.files("/mnt/bohacek/Marrocco_GSE200670_stressATAC/NFbam/", pattern="bam$", full=TRUE)
bams_NF <- bams_NF[1:24]
bams_NF

#create summarized experiment object
#se_NF <- chromVAR::getCounts(alignment_files=bams_NF, 
#                             peaks=merged_peaksM_NF, 
#                             paired =  TRUE, 
#                             format = "bam")

se_NF <- readRDS("/mnt/bohacek/rebecca/training/EpiC/Project/MarroccopeaksNF.SE.rds")


#se_NF
#assay(se_NF)

#plot sequencing depth
#barplot(se_NF$depth)

#adjust the colnames
#colnames(se)
colnames(se_NF) <- gsub("\\.bam","",basename(bams_NF))

#extract group from filename and add as metadata
se_NF$group <- word(colnames(se_NF), 1, sep = "_")
se_NF$group <- as.factor(se_NF$group)
#se_NF$group
#levels(se_NF$group)
se_NF$group<-relevel(se_NF$group, "Control")

#extract sex from filename and add as metadata
se_NF$sex<-word(colnames(se_NF), 2, sep = "_")
se_NF$sex <- as.factor(se_NF$sex)
#se_NF$sex
#levels(se_NF$sex)
se_NF$sex<-relevel(se_NF$sex, "male")

#extract ID from filename and add as metadata
se_NF$id<-word(colnames(se_NF), 3, sep = "_")
se_NF$id <- as.factor(se_NF$id)
#se_NF$id

#give names to regions, by default there are no rownames, use row.names
row.names(se_NF) <- as.character(granges(se_NF))
```

## Data Exploration

Explore data and visualize the difference in the expression profiles using PCA.

```{r, out.width=c('80%'), fig.align='center', fig.cap=c('FIGURE 5: PCA of peak counts for original groups')}
#plot PCA of animals
colors <- c("blue", "darkgreen", "red", "black")
plotMDS(se_NF, col=colors[se_NF$group], labels=se_NF$group, gene.selection="common")
legend("topleft", legend=levels(se_NF$group), col=colors, ncol=2)
```

Also when using only the reads from nucleosome-free regions, there is no clear separation between the groups, therefore, a covariate/surrogate variable will be added (Fig. 5). Differential analysis will directly be performed using the updated model, taking into account the history of stress, acute stress, sex and the covariate. 

```{r}
#add new metadata based on the groups
#se_NF$group

#add history of stress
se_NF$history <- c(rep("history", 6), rep("no", 6), rep("history", 6), rep("no",6))
se_NF$history <- as.factor(se_NF$history)
se_NF$history<-relevel(se_NF$history, "no")
#se_NF$history

#add acute stress
se_NF$stress <- c(rep("no", 12), rep("acute", 12))
se_NF$stress <- as.factor(se_NF$stress)
se_NF$stress<-relevel(se_NF$stress, "no")
#se_NF$stress

#prepare DGE object
y <- DGEList(counts=assays(se_NF)$counts)

#calulate normalizaiton, assumptions are met in this experiment, so we can use TMM normalization
y <- calcNormFactors(y)

#filter by expression
#y <- y[filterByExpr(y,group=se$group),]

#use surrogate 
se_NF <- SEtools::svacor(se_NF, ~ sex + history + stress, n.sv=1)
assays(se_NF) #corrected assay added

#specify model matrix
mm <- model.matrix(~ sex + SV1 + history + stress, data=as.data.frame(colData(se_NF)))
#we use an intercept here, meaning that we compare against the control

#estimate dispersion
y <- estimateDisp(y,mm)

#fit model
fit <- glmQLFit(y, mm)

#check coefficients
colnames(mm)
#[1] "(Intercept)"    "sexfemale"      "SV1"            "historyhistory" "stressacute"

#get results
results_M_NF <- list()

#compare history vs. control
results_M_NF[["Marrocco.history"]] <- glmQLFTest(fit, coef="historyhistory")
topTags(results_M_NF[["Marrocco.history"]])
summary(decideTests(results_M_NF[["Marrocco.history"]]))
#      historyhistory
#Down          1
#NotSig   141225
#Up            9

#compare acute stress vs. control
results_M_NF[["Marrocco.acute"]] <- glmQLFTest(fit, coef="stressacute")
topTags(results_M_NF[["Marrocco.acute"]])
summary(decideTests(results_M_NF[["Marrocco.acute"]]))
#       stressacute
#Down          43
#NotSig    140324
#Up           868

#calculate fdr
for(i in names(results_M_NF)){
  results_M_NF[[i]]$table$FDR <- p.adjust(results_M_NF[[i]]$table$PValue, method = "fdr")
}
```

```{r, out.width=c('80%'), fig.align='center', fig.cap=c('FIGURE 6: Volcanoplot showing significantly accessible nucleosome-free regions between control and acutely-stressed animals')}
#classify sign. up and downregulated peaks
results_M_NF$Marrocco.acute$table$diff[results_M_NF$Marrocco.acute$table$FDR < 0.05 & results_M_NF$Marrocco.acute$table$logFC >0] <- "UP"
results_M_NF$Marrocco.acute$table$diff[results_M_NF$Marrocco.acute$table$FDR < 0.05 & results_M_NF$Marrocco.acute$table$logFC <0] <- "DOWN"

#visualize results
ggplot(results_M_NF$Marrocco.acute$table, aes(logFC, -log10(FDR), col=results_M_NF$Marrocco.acute$table$diff)) + geom_point(size=1) + theme_bw() + scale_color_manual(values = c("blue", "red"), labels =c ("downregulated", "upregulated", "not significant")) + ggtitle('Significantly accessible regions NF: Acute stress') + labs(color = 'DEA, FDR<0.05', x = expression("logFC"), y = expression("-log"[10]*"p-value"))
```

```{r, out.width=c('80%'), fig.align='center', fig.cap=c('FIGURE 7: Heatmap showing differentially-accessible nucleosome-free regions between control and acutely-stressed animals (corrected)')}
#add metadata column to SE object to define colors
metadata(se_NF)$anno_colors$stress <- c(no="#A8BED0", acute="#2F4E67")
metadata(se_NF)$anno_colors$history <- c(no="#F7D9A3", history="#DB9413")
metadata(se_NF)$anno_colors$sex <- c(male="#A87974", female="#E4A6A0")

#visualize uncorrected counts in heatmap
#sechm(se_NF,features = rownames(topTags(results_M_NF$Marrocco.acute, n = Inf, p.value = 0.05)), do.scale=TRUE, assayName="counts", row_title="Stress-responsive regions", gaps_at = "stress", cluster_rows = TRUE, top_annotation=c("sex", "history", "stress"))

#visualize corrected counts in heatmap
sechm(se_NF,features = rownames(topTags(results_M_NF$Marrocco.acute, n = Inf, p.value = 0.05)), do.scale=TRUE, assayName="corrected", row_title="Stress-responsive regions", gaps_at = "stress", cluster_rows = TRUE, top_annotation=c("sex", "history", "stress"))

#save summarized experiments object
saveRDS(se_NF, file="MarroccopeaksNF.SE.rds")
```

911 regions are differentially accessible between acutely stressed (FSS, double-hit) and non-acutely stressed (Control, ARS) animals using only nucleosome-free reads, correcting for the different history of stress, sex and covariate. Again, most of the sign. altered regions (868 regions) show increased accessibility in response to stress (Fig. 6 & 7). History of stress explains only very few sign. different regions (10), which will not be further evaluated here. 


```{r}
#subset all significant stress peaks
#dim(results_M[["Marrocco.acute"]]$table) #99852
#length(merged_peaksM) #99852
peaksM_NF_stress_sig <- merged_peaksM_NF[which(results_M_NF[["Marrocco.acute"]]$table$FDR<0.05)]
#length(peaksM_NF_stress_sig) #911 peaks

#subset sign. upregulated peaks
peaksM_NF_stress_sigup <- merged_peaksM_NF[which((results_M_NF[["Marrocco.acute"]]$table$FDR<0.05) & (results_M_NF[["Marrocco.acute"]]$table$logFC>0))]
#length(peaksM_NF_stress_sigup) #868

#subset sign. downregulated peaks
peaksM_NF_stress_sigdown <- merged_peaksM_NF[which((results_M_NF[["Marrocco.acute"]]$table$FDR<0.05) & (results_M_NF[["Marrocco.acute"]]$table$logFC<0))]
#length(peaksM_NF_stress_sigdown) #43

#save peaks
saveRDS(peaksM_NF_stress_sig, file="peaksM_NF_stress_sig.rds")
saveRDS(peaksM_NF_stress_sigup, file="peaksM_NF_stress_sigup.rds")
saveRDS(peaksM_NF_stress_sigdown, file="peaksM_NF_stress_sigdown.rds")
saveRDS(merged_peaksM, file="merged_peaksM.rds")
```

## Enrichment analysis of stress-upregulated peaks

Enrichment analysis on the NF-peaks was performed using the rGREAT server. As background, all accessible peaks (merged NF peaks file that was previously used to construct the SE object) from the Marrocco dataset will be used. 

```{r, out.width=c('90%'), fig.align='center', fig.cap=c('FIGURE 8: Enrichment analysis of significantly more accessible nucleosome-free regions')}
#perform enrichment analysis using rGREAT server
library(rGREAT)

#adjust seqlevelStyle for rGREAT server
seqlevelsStyle(merged_peaksM_NF) <- "UCSC"
seqlevelsStyle(peaksM_NF_stress_sig) <- "UCSC"
seqlevelsStyle(peaksM_NF_stress_sigup) <- "UCSC"
seqlevelsStyle(peaksM_NF_stress_sigdown) <- "UCSC"

#only keep common chromosome names and clean up Granges for rGREAT
seqnames(merged_peaksM_NF)
chrom <- c("chr1", "chr10", "chr11", "chr12", "chr13", "chr14", "chr15", "chr16", "chr17", "chr18", "chr19", "chr2", "chr3", "chr4", "chr5", "chr6", "chr7", "chr8", "chr9", "chrM", "chrX", "chrY")
merged_peaksM_NF_chrom<- keepSeqlevels(merged_peaksM_NF, value=chrom, pruning.mode="coarse")
length(merged_peaksM_NF_chrom) #141203
names(merged_peaksM_NF_chrom) <-NULL
seqnames(peaksM_NF_stress_sig)
peaksM_NF_stress_sig_chrom<- keepSeqlevels(peaksM_NF_stress_sig, value=chrom, pruning.mode="coarse")
length(peaksM_NF_stress_sig_chrom) #911
names(peaksM_NF_stress_sig_chrom) <-NULL
peaksM_NF_stress_sigup_chrom<- keepSeqlevels(peaksM_NF_stress_sigup, value=chrom, pruning.mode="coarse")
length(peaksM_NF_stress_sigup_chrom) #868
names(peaksM_NF_stress_sigup_chrom) <-NULL
peaksM_NF_stress_sigdown_chrom<- keepSeqlevels(peaksM_NF_stress_sigdown, value=chrom, pruning.mode="coarse")
length(peaksM_NF_stress_sigdown_chrom) #43
names(peaksM_NF_stress_sigdown_chrom) <-NULL

#get enrichment results for all sign peaks
#job_stressNF_sig <- rGREAT::submitGreatJob(gr=peaksM_NF_stress_sig_chrom, bg=merged_peaksM_NF_chrom, species="mm10") #we use mm10, cause GRCm38 is not supported
#res_stressNF_sig <- rGREAT::getEnrichmentTables(job_stressNF_sig)
#res_stressNF_sig

#plot the top 30 Biological processes
#bp_stressNF_sig <- res_stressNF_sig$`GO Biological Process`
#bp_stressNF_sig
#sum(bp_stressNF_sig$Hyper_Adjp_BH<0.05) #49 GO terms are sign. enriched
#ggplot(head(bp_stressNF_sig,30), aes(Hyper_Fold_Enrichment, reorder(name, Hyper_Adjp_BH), size=Hyper_Foreground_Region_Hits, color=-log10(Hyper_Adjp_BH))) + geom_point() + scale_color_viridis_c() + ggtitle("Stress-responsive peaks: Biological Processes") + scale_y_discrete(labels = label_wrap(50))

#get enrichment results for all sign upregulated peaks
job_stressNF_sigup <- rGREAT::submitGreatJob(gr=peaksM_NF_stress_sigup_chrom, bg=merged_peaksM_NF_chrom, species="mm10") #we use mm10, cause GRCm38 is not supported
res_stressNF_sigup <- rGREAT::getEnrichmentTables(job_stressNF_sigup)
#res_stressNF_sigup

#plot the top 30 Biological processes
bp_stressNF_sigup <- res_stressNF_sigup$`GO Biological Process`
#bp_stressNF_sig
sum(bp_stressNF_sigup$Hyper_Adjp_BH<0.05) #52 GO terms are sign. enriched
ggplot(head(bp_stressNF_sigup,30), aes(Hyper_Fold_Enrichment, reorder(name, Hyper_Adjp_BH), size=Hyper_Foreground_Region_Hits, color=-log10(Hyper_Adjp_BH))) + geom_point() + scale_color_viridis_c() + ggtitle("Stress-responsive peaks: Biological Processes") + scale_y_discrete(labels = label_wrap(50))

#get enrichment results for all sign downregulated peaks
job_stressNF_sigdown <- rGREAT::submitGreatJob(gr=peaksM_NF_stress_sigdown_chrom, bg=merged_peaksM_NF_chrom, species="mm10") #we use mm10, cause GRCm38 is not supported
res_stressNF_sigdown <- rGREAT::getEnrichmentTables(job_stressNF_sigdown)
#res_stressNF_sigdown

#plot the top 20 Biological processes
bp_stressNF_sigdown <- res_stressNF_sigdown$`GO Biological Process`
#bp_stressNF_sigdown
sum(bp_stressNF_sigdown$Hyper_Adjp_BH<0.05) #0 GO terms are sign. enriched
```

Performing enrichment analysis on all sign. different more accessible regions between acutely stressed and control animals in the Marrocco dataset using only NF reads, reveals the same enrichment of response to glucocorticoids (Fig. 8). This is consistently observed when testing all sign. altered regions. Sign. downregulated regions however, do not yield any significant enrichment. 

## Compare peaks

Next, we compare the sign. altered peaks between all accessible regions and accessible regions from the nucleosome-free reads only, defined by the previous differential analysis.

```{r, out.width=c('100%'), fig.align='center', fig.cap=c('FIGURE 9: Upset plot showing overlaps of peaks between the regions with significantly altered regions')}
#prepare list of peaks
peaksMarrocco_sig<- GRangesList("ATAC-stress-up" = peaksM_stress_sigup_chrom, "ATAC-stressNF-down" = peaksM_NF_stress_sigdown_chrom, "ATAC-stressNF-up" = peaksM_NF_stress_sigup_chrom)
#peaksMarrocco_sig
regionUpset(peaksMarrocco_sig)
#calculate fraction of shared peaks between the upregulated peaks
sum(overlapsAny(peaksM_stress_sigup_chrom, peaksM_NF_stress_sigup_chrom)) #286
```

Most of the sign. altered upregulated regions of all accessible regions overlap with sign. more accessible peaks of nucleosome-free regions (Fig. 9).

Given the further focus on GR, which is a TF, the nucleosome-free peaks and the corresponding significant peaks will be used for all further analyses.

